[Feature] - Exercise statistics

FuchsDominik opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We want to have specific exercise related statistics

### Description
<!-- Describe your changes in detail -->
Added doughnut charts in the exercise detail page and also in a new exercise statistics page, together with other metrics interesting for exercises.
We display the score distribution, the submissions, active users, active tutors, created results, created feedback

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate into the detail page of an exercise without an releaseDate or with an releaseDate in the future
3. Verify that no doughnut charts are shown and everything looks as usual
4. Move into a exercise which has a already passed releaseDate
5. Make sure the no initial error is thrown
6. Verify that the values in the doughnut charts are correct
6.1 Average score -> go to course statistics and compare with the bar in the average score bar chart
6.2 participations -> click on the title or the doughnut chart and compare the number of participations
6.3 questions -> click on the title or the doughnut chart and count the number of questions the exercise has and how many of them have an answer which is approved 
7. Click on the Statistics button on top of the detail page
8. make sure the values of the doughnut charts are the same as before
9. compare the score distribution graph with the actual scores achieved by the students (on the scores page, accessable by clicking onto the average score doughnut chart title or graph)
10. participate in the exercise and make sure the corresponding graphs graphs behave accordingly (submit something -> bar in the submission chart increases accordingly)
11. Change the tabs (to DAY, MONTH, QUARTER or YEAR) and make sure values and labels are still correct (The doughnut charts and the score distribution graph do not change as those do not depend on a time constraint)

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
StatisticsResource.java: 90%
StatisticsService.java: 94%
StatisticsRepository.java: 86%

doughnut-chart.component.ts: 89%
exercise-statistics.component.ts: 100%
exercise-detail-statistics.component.ts: 100%
(Improved:
text-exercise-detail.component.ts: 98%
modeling-exercise-detail.component.ts: 89%
file-upload-exercise-detail.component.ts: 93%
programming-exercise-detail.component.ts: 57%
)

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

The new Exercise Statistics:
<img width="1439" alt="Screenshot 2021-05-23 at 09 59 58" src="https://user-images.githubusercontent.com/44401560/119253293-cc3b2b00-bbb0-11eb-920b-95fa21e666c4.png">

3 new doughnut charts in the exercise detail view:
<img width="1439" alt="Screenshot 2021-05-23 at 10 41 14" src="https://user-images.githubusercontent.com/44401560/119253891-86339680-bbb3-11eb-8ff6-59cfe3242edb.png">

ivohashamov reviewed (approved):
Tested on TS1, works as described.

KiraNickel reviewed (dismissed):
Tested on ts3, everything works as expected 👍🏼

MaisiKoleni commented:
Nice feature for the statistics page! I am a bit surprised about the inclusion in the exercise details page though, especially in that size at a very prominent place.
Is this always shown or only for released exercises?
Reason is we use the detail page a lot to e.g.  proofread the problem statement, solution and correction or to check the results of template and solution for programming exercises.

From a personal point of view (talk with Stephan about that, only he can decide on that) I would make it at least smaller for the details page to fit into the rest compactness-wise, e.g. a single line or two. Or only activate it for released exercises, because there is no interesting information before that date. (Or in the extreme case leave it in the statistics page only, which is one click away).

cremertim reviewed (approved):
worked on ts1 as expected

maximilianbergmann reviewed (dismissed):
Tested on TS3. Works as expected.

FuchsDominik commented:
> Nice feature for the statistics page! I am a bit surprised about the inclusion in the exercise details page though, especially in that size at a very prominent place.
> Is this always shown or only for released exercises?
> Reason is we use the detail page a lot to e.g. proofread the problem statement, solution and correction or to check the results of template and solution for programming exercises.
> 
> From a personal point of view (talk with Stephan about that, only he can decide) I would make it at least smaller for the details page to fit into the rest compactness-wise, e.g. a single line or two. Or only activate it for released exercises, because there is no interesting information before that date. (Or in the extreme case leave it in the statistics page only, which is one click away).

Hey @MaisiKoleni , thanks for the feedback. You are right that the charts take a lot of space and are not needed for exercises in the preparation phase (release date in the future). I adapted the functionality so those doughnut charts are only shown when the release date has passed as only then the statistics are actually useful. And i made the doughnut charts smaller as well, so they don't take that much space 👍

ndricimrr commented on src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-detail.component.html:
```diff
@@ -12,6 +12,53 @@ <h2><span jhiTranslate="artemisApp.fileUploadExercise.detail.title">File Upload
+                </a>
+            </div>
+            <div *ngIf="fileUploadExercise.releaseDate != undefined && moment(fileUploadExercise.releaseDate).isBefore(moment())">
+                <jhi-doughnut-chart
+                    class="mt-3 d-inline-flex"
+                    [courseId]="fileUploadExercise.course?.id!"
+                    [contentType]="DoughnutChartType.AVERAGE_EXERCISE_SCORE"
+                    [exerciseId]="fileUploadExercise.id!"
+                    [exerciseType]="ExerciseType.TEXT"
+                    [currentPercentage]="doughnutStats?.averageScoreOfExercise"
+                    [currentAbsolute]="absoluteAveragePoints"
+                    [currentMax]="doughnutStats?.maxPointsOfExercise"
+                >
+                </jhi-doughnut-chart>
+                <jhi-doughnut-chart
+                    class="mt-3 d-inline-flex"
+                    [courseId]="fileUploadExercise.course?.id!"
+                    [contentType]="DoughnutChartType.PARTICIPATIONS"
+                    [exerciseId]="fileUploadExercise.id!"
+                    [exerciseType]="ExerciseType.TEXT"
+                    [currentPercentage]="participationsInPercent"
+                    [currentAbsolute]="doughnutStats?.numberOfParticipations"
+                    [currentMax]="doughnutStats?.numberOfStudentsInCourse"
+                >
+                </jhi-doughnut-chart>
+                <jhi-doughnut-chart
+                    class="mt-3 d-inline-flex"
+                    [courseId]="fileUploadExercise.course?.id!"
+                    [contentType]="DoughnutChartType.QUESTIONS"
+                    [exerciseId]="fileUploadExercise.id!"
+                    [exerciseType]="ExerciseType.TEXT"
+                    [currentPercentage]="questionsAnsweredInPercent"
+                    [currentAbsolute]="doughnutStats?.numberOfAnsweredQuestions"
+                    [currentMax]="doughnutStats?.numberOfQuestions"
+                >
```
This part about 3 doughnut - **jhi-doughnut-chart** s  is repeated many times for different types of exercises. It looks like it could be reused. I'm wondering what you think about making this chunk a modular/common reusable component containing the 3 charts.

ndricimrr reviewed (commented):
- Added a small suggestion/question.

- Only a pretty minor thing I saw, in the table graph the score overlaps with the legend if score is at maximum. I'm not sure if this comes with chart.js or if it is even worth it to be fixed. 
![cc02](https://user-images.githubusercontent.com/18335489/119276080-1821ba80-bc19-11eb-96c3-273ef8b20c19.PNG)

Other than that, Tested on TS1, everything else works as expected.

canyelten reviewed (dismissed):
Tested on TS1. Works as expected

FuchsDominik commented:
Hey @ndricimrr , thanks for testing. That's an interesting idea. This definitely decreases duplicity in the code 👍  I moved all this stuff in 76fd28e and also fixed the chart legend problem in 91f563a by moving the legend to the bottom, think that should do the trick

ndricimrr commented:
Tested again on TS1. Issue is fixed 👍

argertboja commented:
Tested on TS1, worked as expected! Nice job 👍🏼

KiraNickel reviewed (approved):
Tested on ts3. Everything works as expected, charts are only shown for released exercises and the smaller version looks really good! 👍🏼  I like it a lot :)

DanielKainz reviewed (approved):
Tested on TS1. Everything works as expected. :-)

iozbeyli commented on src/main/webapp/app/shared/statistics-graph/statistics-score-distribution-graph.component.ts:
```diff
@@ -0,0 +1,135 @@
+    @Input()
+    numberOfExerciseScores: number | undefined;
+
+    // Html properties
+    LEFT = false;
+    RIGHT = true;
+    SpanType = SpanType;
```
Where are these used? I cannot see them in the template

FuchsDominik replied:
True, I kept this variable which is not used anymore. Removed it

iozbeyli commented on src/main/webapp/app/exercises/shared/statistics/exercise-statistics.component.html:
```diff
@@ -0,0 +1,34 @@
+        [scoreDistribution]="exerciseStatistics?.scoreDistribution"
+        [numberOfExerciseScores]="exerciseStatistics?.numberOfExerciseScores"
+    ></jhi-statistics-score-distribution-graph>
+    <div *ngFor="let graph of graphTypes; let i = index" id="graph-{{ i }}" class="graph">
```
Id is a bit too generic, can be a problem in the future.

FuchsDominik replied:
👍

iozbeyli commented on src/main/webapp/app/exercises/shared/statistics/exercise-management-statistics-dto.ts:
```diff
@@ -0,0 +1,17 @@
+    numberOfQuestions: number;
+    numberOfAnsweredQuestions: number;
+
+    constructor() {}
```
I think this is the default value so you can omit this.

FuchsDominik replied:
👍

iozbeyli commented on src/main/webapp/app/exercises/shared/statistics/doughnut-chart.component.ts:
```diff
@@ -0,0 +1,100 @@
+            backgroundColor: 'rgba(0, 0, 0, 1)',
+            callbacks: {
+                label(tooltipItem: any, data: any) {
+                    const value = data['datasets'][0]['data'][tooltipItem['index']];
```
This just looks like a ticking time bomb 😄 One undefined and boom 💥

FuchsDominik replied:
Agree on the 💥 , I took it from the internet and it worked until now, but I added an additional check for safety reasons now :) 0c371b3

iozbeyli commented on src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-detail.component.ts:
```diff
@@ -53,6 +68,12 @@ export class FileUploadExerciseDetailComponent implements OnInit, OnDestroy {
                 },
                 (res: HttpErrorResponse) => this.onError(res),
             );
+        this.statisticsService.getExerciseStatistics(exerciseId).subscribe((statistics: ExerciseManagementStatisticsDto) => {
+            this.doughnutStats = statistics;
+            this.participationsInPercent = statistics.numberOfStudentsInCourse > 0 ? round((statistics.numberOfParticipations / statistics.numberOfStudentsInCourse) * 100, 1) : 0;
+            this.questionsAnsweredInPercent = statistics.numberOfQuestions > 0 ? round((statistics.numberOfAnsweredQuestions / statistics.numberOfQuestions) * 100, 1) : 0;
+            this.absoluteAveragePoints = round((statistics.averageScoreOfExercise * statistics.maxPointsOfExercise) / 100, 1);
+        });
```
You are doing the same operation whenever you call this service method, maybe you should return them from the service as well.

ArtemTsurikov reviewed (approved):
Tested on TS1, looks good!

MDK5 reviewed (approved):
Tested in testing session on ts1, everything worked

coolchock reviewed (approved):
tested in testing session, everything works as expected

tobiasjungmann reviewed (approved):
tested in the testing session. Worked as described.

jhainzlschmid reviewed (approved):
Tested in testing session. Works as expected

cremertim reviewed (approved):
tested on ts3. Works as expected!

NicolasUllmann reviewed (approved):
Tested on Server 1. Everything worked as expected.

willergeorg reviewed (approved):
Tested on TS1. Worked fine.
